packages/graphql/src/services/practiceQuizzes.ts

```diff
@@ -27,32 +39,247 @@ export async function getPracticeQuizData(
+      if (aResponse.correctCountStreak < bResponse.correctCountStreak) return -1
+      if (aResponse.correctCountStreak > bResponse.correctCountStreak) return 1
+      if (aResponse.correctCount < bResponse.correctCount) return -1
+      if (aResponse.correctCount > bResponse.correctCount) return 1
+      if (aResponse.lastCorrectAt < bResponse.lastCorrectAt) return -1
+      if (aResponse.lastCorrectAt > bResponse.lastCorrectAt) return 1
+      return 0
+    })
+
+    return {
+      ...quiz,
+      stacks: orderedStacks,
+    }
+  }
   return quiz
 }
+
+interface RespondToFlashcardInstanceInput {
+  id: number
+  courseId: string
+  correctness: number
+}
+
+export async function respondToFlashcardInstance(
+  { id, courseId, correctness }: RespondToFlashcardInstanceInput,
+  ctx: Context
+) {
+  const existingInstance = await ctx.prisma.elementInstance.findUnique({
+    where: {
+      id,
+    },
+  })
+
+  if (!existingInstance) return null
+
+  // 0 = wrong, 1 = partial, 2 = correct
+  if (![0, 1, 2].includes(correctness)) {
+    return null
+  }
+
+  // TODO: also handle the case where the participant is logged in, but has no valid participation in the considered course
+  if (!ctx.user?.sub) {
+    // TODO: if user is not logged in, do not check the questionResponse, only save it on the instance / update the instance results
+    return null
+  }
+
+  // update the aggregated data on the element instance
+  await ctx.prisma.elementInstance.update({
+    where: {
+      id,
+    },
+    data: {
+      results: {
+        wrong: existingInstance.results.wrong + (correctness === 0 ? 1 : 0),
+        partial: existingInstance.results.partial + (correctness === 1 ? 1 : 0),
+        correct: existingInstance.results.correct + (correctness === 2 ? 1 : 0),
+        total: existingInstance.results.total + 1,
+      },
+    },
+  })
+
+  // find existing question response to this instance by this user
+  const existingResponse = await ctx.prisma.questionResponse.findUnique({
+    where: {
+      participantId_elementInstanceId: {
+        participantId: ctx.user.sub,
+        elementInstanceId: id,
+      },
+    },
+  })
+
+  ctx.prisma.questionResponseDetail.create({
+    data: {
+      response: {
+        correctness: correctness,
+      },
+      participant: {
+        connect: { id: ctx.user.sub },
+      },
+      questionInstance: {
+        connect: { id },
+      },
+      participation: {
+        connect: {
+          courseId_participantId: {
+            courseId,
+            participantId: ctx.user.sub,
+          },
+        },
+      },
+    },
+  })
+
+  // TODO: think about possibility to fuse these two cases with a single upsert (however, keep in mind that we need to update JSON stuff here)
+  if (existingResponse) {
+    let aggregatedResponse =
+      existingResponse.aggregatedResponses as AggregatedResponseFlashcard
+    switch (correctness) {
+      case 0:
```

Comment:
if we use an enum and not ints but words for the keys, we could do it generically here and increment using 

...aggregatedResponse,
[response]: aggregatedResponse[response] + 1,
total: aggregatedResponse[response] + 1

and just check precedingly whether the response is one of the three in the enum

Reply:
wip implementation has been committed